### PR TITLE
unixtransport: add Install hook for HTTP(S)+UNIX transport over httptest.Server

### LINKED ISSUE #2: Support for an analogous API to httptest.NewServer
I'm working a project where I want to support both HTTP and UNIX+HTTP transports, so I'm making use of the functionality here on the client side.

For server testing, I came up with the following:

```go
// testUNIX creates a *zedhook.Client backed by a UNIX socket HTTP server which
// returns its payload on a channel.
func testUNIX(t *testing.T) (*zedhook.Client, <-chan zedhook.Payload) {
	t.Helper()

	pC := make(chan zedhook.Payload, 1)
	srv := unixtestNewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var p zedhook.Payload
		if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
			panicf("failed to decode JSON: %v", err)
		}

		pC <- p
		w.WriteHeader(http.StatusNoContent)
	}))
	t.Cleanup(srv.Close)

	c, err := zedhook.NewClient(srv.URL+"/push", nil)
	if err != nil {
		t.Fatalf("failed to create zedhook client: %v", err)
	}

	return c, pC
}
```

The API mimics the one exposed by `httptest.NewServer`. Would you accept a PR to add `unixtransport.TestServer` or similar? Thanks again!

### PR DESCRIPTION
**Pull request by mdlayher**
Signed-off-by: Matt Layher <mdlayher@gmail.com>

I started with just the basics I anticipate my project will need. Feedback very welcome.

Fixes #2.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// NewTestServer starts and returns a new TestServer. The caller should call
+// Close when finished, to shut it down.
+func NewTestServer(handler http.Handler) *TestServer {
+	// If any subsequent calls fail before we return *TestServer, be sure to
+	// os.RemoveAll(dir) to tidy things up.
+	dir, err := os.MkdirTemp(os.TempDir(), "unixtransport-server")
+	if err != nil {
+		panicf("failed to make temporary directory: %v", err)
```
Comment (peterbourgon) [inline server.go:34]: ...this to `tb.Fatal` instead of panicking?

File: server.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
```
Comment (peterbourgon) [inline server.go:36]: `defer os.RemoveAll(...)`?

File: server.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	server := httptest.NewUnstartedServer(handler)
+	server.Listener = l
+	server.Start()
+
+	// And finally create an http.Client which can speak http+unix scheme.
+	client := server.Client()
+	Register(client.Transport.(*http.Transport))
+
+	return &TestServer{
+		URL: fmt.Sprintf("http+unix://%s:", l.Addr().String()),
```
Comment (peterbourgon) [inline server.go:53]: Is this correct? The colon seems suspect?

File: server.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package unixtransport
```
Comment (peterbourgon) [inline server.go:1]: Should this be a separate package, e.g. `unixtransport/unixtransporttest`? Honest question, don't actually know 😅

File: server.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+)
+
+// A TestServer is an HTTP+UNIX server listening on a temporary UNIX socket, for
```
Comment (peterbourgon) [inline server.go:12]:
```suggestion
// TestServer is an HTTP+UNIX server listening on a temporary UNIX socket, for
```

**Review by peterbourgon — COMMENTED** [review]
Just a bunch of questions for now...

File: server.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// use in end-to-end HTTP tests. TestServer's API mimics that of the
+// net/http/httptest.Server type.
+type TestServer struct {
+	// Base URL of form http+unix:///tmp/(random)/test.sock: with no trailing
+	// slash.
+	URL string
```
[the remark this answers is not in the document]
Comment (mdlayher) [reply server.go:18]: Yep, the API mimics that of httptest.Server:

```go
	srv := unixtransport.NewTestServer(handler)
	t.Cleanup(srv.Close)

	c, err := NewClient(srv.URL+"/push", srv.Client())
	if err != nil {
		t.Fatalf("failed to create UNIX+HTTP zedhook client: %v", err)
	}
	```

File: server.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	server *httptest.Server
+	client *http.Client
+
+	// Temporary directory, cleaned up on close.
+	dir string
+}
+
+// NewTestServer starts and returns a new TestServer. The caller should call
+// Close when finished, to shut it down.
+func NewTestServer(handler http.Handler) *TestServer {
```
[the remark this answers is not in the document]
Comment (mdlayher) [reply server.go:29]: Yeah it definitely can. I mimicked the stdlib `httptest.NewServer`. Originally my version took a `testing.TB`, but I wanted to keep it as close as possible to stdlib.

Though I also think taking a `tb` is better, so I'll update.

File: server.go (hunk 1 above)
Comment (mdlayher) [reply server.go:34]: Ack.

File: server.go (hunk 3 above)
Comment (mdlayher) [reply server.go:53]: Yep, it is. I am using the format you documented for compatibility with the modified transport:
```
=== RUN   TestClientPush/UNIX
(string) (len=57) "http+unix:///tmp/unixtransport-server815910969/test.sock:"
(string) (len=62) "http+unix:///tmp/unixtransport-server815910969/test.sock:/push"
```

The first is raw `srv.URL`, the second is with my specific handler path.

File: server.go (hunk 4 above)
Comment (mdlayher) [reply server.go:1]: Up to you. Personally I think that since both the existing and new APIs are so small, it's probably fine to keep it all in one package.

That said I have no qualms with choosing to revisit this later and won't mind updating my code if we decide to move this.

File: server.go (hunk 2 above)
Comment (mdlayher) [reply server.go:36]: I don't believe so: right now we generate the temporary directory and either remove it only on error returns, or on a call to Close. I don't know what happens if we have a UNIX socket listener open but then remove the directory it exists in before returning to the caller.

If we change the API to take a `testing.TB` though, I can use `tb.TempDir()` and let the test clean it all up at the end. That is simpler and easier to reason about too.

**Comment by mdlayher** [comment]
Updated! Happy to squash once you're all happy, or we can go with GitHub squash and merge. But please let me know if you have any further questions or feedback, thanks!

File: server.go (hunk 3 above)
Comment (mdlayher) [reply server.go:44]: Though I guess the question is: should I be required to add the `:/push` suffix in my own code?

Comment (peterbourgon) [reply server.go:44]: Ah, I see what you're doing here. I think this is a reasonable approach. The more formal method would be for the package to offer some URL-building functionality, but I'm not sure that juice is worth the squeeze.

**Comment by peterbourgon** [comment]
So I'm totally down with the thing you're trying to do here, and I'm happy to add capabilities to help you do it. But there's something about this approach which feels a bit off to me. This package is just a single exported function, which  interacts with stuff at the transport layer, somewhere underneath the HTTP clients themselves. What you want is some helpers to stand up the other half of that communication path, but I think that's not the HTTP server itself, I think it's at a lower layer, symmetric with the client's transport. 

That is, I think it should be a, like, TestListener? which you would assign to the stock httptest.Server's Listener field? That's not enough by itself, you'd also need the Client method to return something that could speak to it. But I think this is possible, and would be about the same work for you to use in a test. Let me poke at it.

**Comment by peterbourgon** [comment]
I mean this is basically what I'm doing at the top of TestBasics, just generalized a bit. I think it'll work.

**Comment by peterbourgon** [comment]
I think it might just be this? Am I missing something? I'm probably missing something.

```go
func NewTestServer(tb testing.TB) *httptest.Server {
	tb.Helper()

	var s httptest.Server

	ln, err := net.Listen("unix", filepath.Join(tb.TempDir(), "unixtransport"))
	if err != nil {
		tb.Errorf("unixtransport: httptest.Server not configured: %v", err)
		return &s
	}

	tb.Cleanup(func() {
		if err := ln.Close(); err != nil {
			tb.Errorf("unixtransport: close listener: %v", err)
		}
	})

	s.Listener = ln
	Register(s.Client().Transport.(*http.Transport))
	return &s
}
```

**Comment by peterbourgon** [comment]
Maybe it would be better to offer something like `func Install(tb testing.TB, s *httptest.Server)` which would set `s.Listener`, Register in the client's transport, and set `s.URL` based on the state of `s.TLS`. That would preserve all of the original behavior, I think?

**Comment by mdlayher** [comment]
Thanks for the ideas! I'll be going out of town for a couple of days but will try to play with this more and see what I can come up with. I like the Install idea and wonder if that could be workable if the Server is forcibly started and the URL is then replaced with one that complies with the expected format here.

**Comment by mdlayher** [comment]
I spiked out a version of `unixtransport.Install` that should work for my needs. Thoughts on https://github.com/mdlayher/zedhook/commit/3d9f21af4b0e620114478f39aa1e0f2b22f381f0?

**Comment by peterbourgon** [comment]
Ah, you call `Start`, which solves the problem. Yeah, I'd want to workshop the name a bit more, let me noodle on it, but I think this is a reasonable approach.

**Comment by mdlayher** [comment]
Sounds good. I'll port that code over here and also call StartTLS in place of Start when needed.

**Comment by mdlayher** [comment]
Updated. I didn't add a TLS-specific test for this case but could duplicate the same structure as TestBasics if you'd like.

**Comment by peterbourgon** [comment]
I think this is the right code, but I'm still noodling on the right way to express it. Sorry for the delay.

**Comment by mdlayher** [comment]
Not a problem!

**Comment by mdlayher** [comment]
No longer needed!
